extension/js/common/composer/interfaces/composer-types.ts

```diff
@@ -2,12 +2,18 @@
 
 'use strict';
 
-export type RecipientElement = {
-  email: string,
-  element: HTMLElement
+export type SendingType = 'to' | 'cc' | 'bcc';
+
+export interface BaseRecipient {
```

Comment:
Why not `Recipients = {to: string[], cc?: string[], bcc?: string[]}` ? That will simplify your code

Reply:
Sorry, but I don't exactly understand how it will simplify the code.

Could you please explain to me a little bit more what do you mean?

Reply:
You will be able to remove this whole chunk of code:
```js
    const to: string[] = [];
    const cc: string[] = [];
    const bcc: string[] = [];
    for (const recipient of recipients) { // Using JS loop to not dublicate .filter().map()
      switch (recipient.sendingType) {
        case 'to':
          to.push(recipient.email);
          break;
        case 'cc':
          cc.push(recipient.email);
          break;
        case 'bcc':
          bcc.push(recipient.email);
          break;
      }
    }
```
createMsgObj API will become more obvious, from:
```js
await Google.createMsgObj(acctEmail, acctEmail, [{ email: acctEmail, sendingType: 'to' }], subj, { 'text/html': emailMsg }, emailAtts);
```
to:
```js
await Google.createMsgObj(acctEmail, acctEmail, { to: [acctEmail] }, subj, { 'text/html': emailMsg }, emailAtts);
```

Reply:
Okay, got it,
If it's the only reason so I can just change the code this way:
```TS
    const result: Recipients = { to: [], cc: [], bcc: [] };
    // tslint:disable-next-line: standard-loops
    recipients.forEach(r => result[r.sendingType].push(r.email));
```

Because there are some places where it's better to use `BaseRecipient` or `RecipientElement` like here:

if we change method signature from this:
```TS
  public static createMsgObj = async (
    acctEmail: string, from: string = '', recipientElements: RecipientElement[] = [], subject: string = '', by: SendableMsgBody, atts?: Att[], threadRef?: string
  ): Promise<SendableMsg> 
```
to that:
```TS
  public static createMsgObj = async (
    acctEmail: string, from: string = '', recipients: Recipients, subject: string = '', by: SendableMsgBody, atts?: Att[], threadRef?: string
  ): Promise<SendableMsg> 
```
Then we need to join (`Array.join`) all `to:`, `cc:` and `bcc:` values to make this validation working. (I guess I will need to do these joins in other places too).
```TS
    if (!recipientElements.length) {
      throw new Error('The To: field is empty. Please add recipients and try again');
    }
    const invalidEmails = recipientElements.filter(r => !Str.isEmailValid(r.email));
    if (invalidEmails.length) {
      throw new Error(`The To: field contains invalid emails: ${invalidEmails.join(', ')}\n\nPlease check recipients and try again.`);
    }
```

So, I would like to leave it as is but change that large `for` cycle to the small one I have written above.

Reply:
I'll take a look later today

Reply:
In `if (!recipientElements.length) {` you should actually do `recipients.to.length`. Email without `To` header is invalid (to my knowledge). So you would, as it is now, have to rewrite that with a filter anyway.

Maybe you meant `Array.concat` ? You can do `Array.prototype.concat(...Object.values(recipients))`. If there are more than two places that need this, you could put it as ComposerRecipients (or whatever is the class name) method, etiher normal or static, eg `asArray(recipients: Recipients): string[]` or something like that.

I didn't examine the code in deep detail so I may be missing something. I just feel like `(acctEmail, acctEmail, [{ email: acctEmail, sendingType: 'to' }]` is a bit awkward as an API.

Reply:
Okay, got it, 
Yes, it makes sense, I will rewrite it according to what you just wrote.

Reply:
It should be easier to use  :+1:

Reply:
That's always nice :)

Reply:
It isn't resolved yet

I will resolve it when I finish all other things